File: algorithms/algorithms-cplusplus/leetcode/Question_0648_Replace_Words.cc
#include <string>
#include <vector>
#include <sstream>
#include <iostream>

using namespace std;

struct TrieNode {
  TrieNode* children[26] = {};
  bool isWord = false;
};

class Solution {
 public:
  string replaceWords(vector<string>& dictionary, string sentence) {
    TrieNode root;
    for (string s : dictionary) {
      TrieNode* cur = &root;
      for (char c : s) {
        if (!cur->children[c - 'a']) {
          cur->children[c - 'a'] = new TrieNode();
        }
        cur = cur->children[c - 'a'];
      }
      cur->isWord = true;
    }
    string res = "";
    string token;
    stringstream ss(sentence);
    while (getline(ss, token, ' ')) {
      TrieNode* cur = &root;
      int i = 0;
      for (; i < token.size(); i++) {
        if (!cur->children[token[i] - 'a'] || cur->isWord) {
          break;
        }
        cur = cur->children[token[i] - 'a'];
      }
      res += cur->isWord ? token.substr(0, i) : token;
      res += " ";
    }
    return res.substr(0, res.size() - 1);
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0648_Replace_Words.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.List;

public class Question_0648_Replace_Words {
    class TrieNode {
        TrieNode[] children;
        String word;

        TrieNode() {
            children = new TrieNode[26];
        }
    }

    public String replaceWords(List<String> dictionary, String sentence) {
        TrieNode trie = new TrieNode();
        for (String s : dictionary) {
            TrieNode cur = trie;
            for (char c : s.toCharArray()) {
                if (cur.children[c - 'a'] == null) {
                    cur.children[c - 'a'] = new TrieNode();
                }
                cur = cur.children[c - 'a'];
            }
            cur.word = s;
        }
        StringBuilder res = new StringBuilder();
        for (String word : sentence.split("\\s+")) {
            if (res.length() > 0) {
                res.append(" ");
            }
            TrieNode cur = trie;
            for (char c : word.toCharArray()) {
                if (cur.children[c - 'a'] == null || cur.word != null) {
                    break;
                }
                cur = cur.children[c - 'a'];
            }
            res.append(cur.word != null ? cur.word : word);
        }
        return res.toString();
    }
}
